src/pool.ts

```diff
@@ -0,0 +1,729 @@
+      );
+    }
+
+    this._proxy = proxy;
+  }
+
+  private async _release(): Promise<void> {
+    // Release this connection holder.
+    if (this._inUse === null) {
+      // The holder is not checked out.
+      return;
+    }
+
+    if (!this._inUse.done) {
+      await this._inUse.setResult();
+    }
+
+    this._inUse = null;
+
+    // Deinitialize the connection proxy.  All subsequent
+    // operations on it will fail.
+    if (this._proxy !== null) {
+      this._proxy.detach();
+      this._proxy = null;
+    }
+
+    // Put ourselves back to the pool queue.
+    this._pool.queue.push(this);
+  }
+}
+
+export class PoolConnectionProxy {
+  private _holder: PoolConnectionHolder;
+  private _connection: AwaitConnection | null;
+
+  constructor(holder: PoolConnectionHolder, connection: AwaitConnection) {
+    this._holder = holder;
+    this._connection = connection;
+
+    // Registers a callback that will be invoked when the
+    // when the connection is closed, release it
+    connection.setProxy(this);
+  }
+
+  public get connection(): AwaitConnection {
+    if (this._connection === null) {
+      throw new errors.InterfaceError("The proxy is detached");
+    }
+    return this._connection;
+  }
+
+  /**
+   * Returns a value indicating whether this PoolConnectionProxy is detached
+   * from a connection.
+   */
+  public get detached(): boolean {
+    return this._connection === null;
+  }
+
+  /**
+   * Returns a value indicating whether this PoolConnectionProxy is detached
+   * or the underlying connection is closed.
+   */
+  isClosed(): boolean {
+    return this._connection === null || this._connection.isClosed();
+  }
+
+  /** @internal */
+  onConnectionClose(): void {
+    this._holder.releaseOnClose();
+  }
+
+  public get holder(): PoolConnectionHolder {
+    return this._holder;
+  }
+
+  detach(): AwaitConnection | null {
+    if (this._connection === null) {
+      return null;
+    }
+
+    const connection = this._connection;
+    this._connection = null;
+    connection.setProxy(null);
+    return connection;
+  }
+}
+
+const DefaultMinPoolSize = 0;
+const DefaultMaxPoolSize = 100;
+
+export interface PoolOptions {
+  connectOptions: ConnectConfig;
+  minSize?: number;
+  maxSize?: number;
+  onAcquire?: (proxy: PoolConnectionProxy) => Promise<void>;
+  onRelease?: (proxy: PoolConnectionProxy) => Promise<void>;
+  onConnect?: (connection: AwaitConnection) => Promise<void>;
+  connectionFactory?: (options?: ConnectConfig) => Promise<AwaitConnection>;
+}
+
+/**
+ * A connection pool.
+ *
+ * Connection pool can be used to manage a set of connections to the database.
+ * Connections are first acquired from the pool, then used, and then released
+ * back to the pool. Once a connection is released, it's reset to close all
+ * open cursors and other resources *except* prepared statements.
+ * Pools are created by calling :func:`~edgedb.pool.create`.
+ */
+export class Pool {
```

Comment:
Can you also add a `CallbackPool` implementation that would provide a callback based API for non async/await code? See the `client.ts` second `Connection` class.

Reply:
So maybe for consistency, I will rename the current `Pool` into `AwaitPool`, and have the `Pool` class work using callbacks. To follow: `AwaitConnection` - `Connection`; `AwaitPool` - `Pool`.

Reply:
I actually want to rename AwaitConnection to Connection, and Connection to CallbackConnection ;) So I'd keep Pool as is.

Reply:
@1st1 I propose this solution to implement the `CallbackPool`:
instead of having code repetition of a second class that looks almost the same to the existing `Connection` in client.ts, I propose to add a common interface with methods that exist in both `Connection` and `Pool` object, I called it "IConnection"; then a generic class that expects `T extends IConnection`. https://github.com/edgedb/edgedb-js/pull/52/commits/d67f6aff6fb26c3828ea4f4b8c7c83ecab627aa6

This way we don't have code repetition. This approach would be a perfectly legitimate way of working with C#, and TypeScript generics look like ~1:1 replica of C# generics.

Classes defined this way can also be extended if desired, if the constructor is changed to be `protected` instead of `private`. For example:

```ts
export class CallbackPool extends CallbackConnectionBase<Pool> {

  constructor(pool: Pool) {
    super(pool);
  }

  example(callback: NodeCallback<null> | null = null): void {
    this.conn
      .close()
      .then((_value) => (callback ? callback(null, null) : null))
      .catch((error) => (callback ? callback(error, null) : null));
  }
}
```

Reply:
Yeah, that works